### PR TITLE
Toehold for running ingest, upload, notifications in test context 2

### PR DESCRIPTION
**Pull request by kanishka-azimi**
Add toehold tests for ingest, upload, and notifications.

Check the following before merging:
- [x] send one email (register)
- [x] perform an upload

rebuild/run tests once locally.

When test suites are ready, 
- upload will move to unit-impure
- ingest will move to integration
- notifications will move to integration

### CODE REVIEW

File: test/TestHarness.hs [hunk 1]
```diff
@@ -305,6 +322,14 @@ withLogs ctxt = do
     logs <- initLogs mempty
     pure (addCntxt ctxt (blankContext { ctxLogs = Just logs }))
 
+mkLogsStub :: IO Logs
+mkLogsStub = initLogs mempty
+
+mkNotificationsStub :: IO Notifications
+mkNotificationsStub = do
+    conf <- C.load "databrary.conf" -- alternatively, don't use config
+    initNotifications (conf C.! "notification")
```
Comment (kanishka-azimi) [inline test/TestHarness.hs]: abstract pattern

**Comment by kanishka-azimi** [comment]
I need to merge this soon, its getting too big and risking a merge conflict. I'll try to polish up whatever I have for now.

File: test/ModelTest.hs [hunk 2]
```diff
@@ -554,7 +563,117 @@ test_17 = localOption (mkTimeout (10 * 10^(6 :: Int))) $ Test.stepsWithResourceA
...
+        (\n m l ml -> (mkDbContext cn2) {
+                ctxNotifications = Just n
+              , ctxMessages = Just m
+              , ctxLogs = Just l
+              , ctxMailer = Just ml
+              })
+            <$> mkNotificationsStub
+            <*> loadMessages
+            <*> mkLogsStub
+            <*> pure initMailer
```
Comment (kanishka-azimi) [inline test/ModelTest.hs]: I'm probably going to use an IORef [Msg] , although switching from IO monad to MonadIO (or something even lighter?) might allow me to use StateT and just pass thread a state of [Msg]. If you have a good intuition, let me know.

File: test/ModelTest.hs [hunk 3]
```diff
@@ -527,7 +536,7 @@ test_16 = ignoreTest $ -- TODO: enable this inside of nix build with solr binari
 
 -------- ezid --------------
 test_17 :: TestTree
-test_17 = localOption (mkTimeout (10 * 10^(6 :: Int))) $ Test.stepsWithResourceAndTransaction "test_17" $ \step ist cn2 -> do
+test_17 = localOption (mkTimeout (15 * 10^(6 :: Int))) $ Test.stepsWithResourceAndTransaction "test_17" $ \step ist cn2 -> do
```
[the remark this answers is not in the document]
Comment (kanishka-azimi) [reply test/ModelTest.hs]: I have integration tests in the unit suite. I'm waiting for when we create the integration suite, probably soon.

File: test/ModelTest.hs [hunk 4]
```diff
@@ -554,7 +563,117 @@ test_17 = localOption (mkTimeout (10 * 10^(6 :: Int))) $ Test.stepsWithResourceA
     --   example - https://doi.org/10.5072/FK2.801
     pure ()
 
--- remaining complex subsystems to demonstrate using: notifications, upload, ingest
+--------- ingest ------------
+test_18 :: TestTree
+test_18 = localOption (mkTimeout (10 * 10^(6 :: Int))) $ Test.stepsWithTransaction "test_18" $ \step cn2 -> do
```
[the remark this answers is not in the document]
Comment (kanishka-azimi) [reply test/ModelTest.hs]: see above, but generally, the following test siutes, as soon as possible:

unit-pure
unit-impure
integration

with integration and unit-impure moving into a new cabal package that depends on databrary package. 

unit-pure runs all the time
unit-impure and integration run a few times as you are working up towards a merge

File: test/ModelTest.hs [hunk 5]
```diff
@@ -554,7 +563,117 @@ test_17 = localOption (mkTimeout (10 * 10^(6 :: Int))) $ Test.stepsWithResourceA
...
+-- TODO: daily notification logic (cleanNotifications + updateStateNotifications)
+
+--------- upload ------------
+test_20 :: TestTree
+test_20 = localOption (mkTimeout (1 * 10^(6 :: Int))) $ Test.stepsWithTransaction "test_20" $ \step cn2 -> do
+    step "Given an authorized investigator and their volume"
+    (aiAcct, aiCtxt) <- addAuthorizedInvestigatorWithInstitution' cn2
+    vol <- runReaderT (addVolumeWithAccess aiAcct) aiCtxt
+    step "When start upload and send all chunks"
+    aiCtxt2 <- (\e s -> aiCtxt { ctxEntropy = Just e , ctxStorage = Just s })
```
[the remark this answers is not in the document]
Comment (kanishka-azimi) [reply test/ModelTest.hs]: I have debated whether cleanup can happen at then end of all tests or per test. If all temp directories are under one root directory for test run, then we could delete the root directory after a test suite run.

File: test/ModelTest.hs [hunk 6]
```diff
@@ -554,7 +563,117 @@ test_17 = localOption (mkTimeout (10 * 10^(6 :: Int))) $ Test.stepsWithResourceA
...
+    Right _ <- runReaderT (ingestJSON vol updateJson True False) aiCtxt7
+    step "Then the user can view the created session"
+    cntrs <- runReaderT (lookupVolumeContainers vol) aiCtxt -- TODO: extract logic from volumeJSONField "containers"
+    (Just "cont1") `elem` (fmap (containerName . containerRow) cntrs) @? "volume doesn't have container naemd cont1"
+    -- TODO: record; >>> non-AV asset <<< ;
+    --   container linked to record + non-AV asset; >>> AV asset <<< ; >>> container linked to record + AV asset <<<
+    --  NOTE: for each transcoded asset, need to wait and manually call collectTranscode based on log status
+
+mkIngestInput :: T.Text -> Value
+mkIngestInput volName =
```
[the remark this answers is not in the document]
Comment (kanishka-azimi) [reply test/ModelTest.hs]: planned for Model.Factories or Model.Factories.Ingest

File: src/Controller/Upload.hs [hunk 7]
```diff
@@ -54,19 +63,25 @@ uploadStart = action POST (pathJSON >/> pathId </< "upload") $ \vi -> withAuth $
...
-  liftIO $ bracket
-    (openFd file WriteOnly (Just 0o640) defaultFileFlags{ exclusive = True })
-    closeFd
-    (`setFdSize` COff size)
+  tok <- createUploadSetSize vol uploadStartRequest
   return $ okResponse [] $ unId (view tok :: Id Token)
 
+createUploadSetSize :: (MonadHas Entropy c m, MonadDB c m, MonadHasIdentity c m, MonadStorage c m) => Volume -> UploadStartRequest -> m Upload
+createUploadSetSize vol (UploadStartRequest filename size) = do
+    liftIO $ print "creating Upload..." --DEBUG
```
[the remark this answers is not in the document]
Comment (kanishka-azimi) [reply src/Controller/Upload.hs]: This logging was for troubleshooting a critical blocker when trying to merge a huge branch for Obsidian. I'll delete it now, we can take up logging in other modules.

File: src/Controller/Upload.hs [hunk 8]
```diff
@@ -54,19 +63,25 @@ uploadStart = action POST (pathJSON >/> pathId </< "upload") $ \vi -> withAuth $
...
-  liftIO $ print "peeking..." --DEBUG
-  file <- peeks $ uploadFile tok
-  liftIO $ bracket
-    (openFd file WriteOnly (Just 0o640) defaultFileFlags{ exclusive = True })
-    closeFd
-    (`setFdSize` COff size)
+  tok <- createUploadSetSize vol uploadStartRequest
   return $ okResponse [] $ unId (view tok :: Id Token)
 
+createUploadSetSize :: (MonadHas Entropy c m, MonadDB c m, MonadHasIdentity c m, MonadStorage c m) => Volume -> UploadStartRequest -> m Upload
```
[the remark this answers is not in the document]
Comment (kanishka-azimi) [reply src/Controller/Upload.hs]: I don't understand this comment, as I don't see how to make liftio span more statements here

File: test/ModelTest.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (kanishka-azimi) [reply test/ModelTest.hs]: For this test, I would be willing to build out a mock. There are two parts of ezid that I would still like to test against their live test site, so I will keep two versions of this test and move this to the TBD integration test suite.

File: src/Controller/Upload.hs (hunk 8 above)
[the remark this answers is not in the document]
Comment (chreekat) [reply src/Controller/Upload.hs]: Oh, you're right. I just saw all the `liftIO` statements without attending to the fact that they're just debug printing.

File: src/Service/Mail.hs [hunk 9]
```diff
@@ -24,7 +26,24 @@ import Has
...
-type MonadMail c m = (MonadLog c m)
+-- | Make an instance of Mailer with default sendMail implementation, for use in
+-- settings where mailer is not being mocked
+initMailer :: Mailer
+initMailer = Mailer
+    { mlr = sendMailImpl
+    }
+
+-- | Default implementation of sendMail to be used in most environments besides unit tests
+sendMailImpl :: String -> Int -> String -> String ->  Mail -> IO ()
```
[the remark this answers is not in the document]
Comment (kanishka-azimi) [reply src/Service/Mail.hs]: Generally yes, and I implemented the newtypes, but I had not intended to touch Service.Mail at all until mocking became necessary. I would normally do this as a later general refactor, not during this change.

**Comment by kanishka-azimi** [comment]
Criteria for moving to integration test suite (as opposed to unit-impure):

- requires multiple complex dependencies in it's context (e.g. notifications + solr)

OR

- requires longer than 5 seconds to run

File: test/ModelTest.hs [hunk 10]
```diff
@@ -614,35 +626,149 @@ test_16 = ignoreTest $ -- TODO: enable this inside of nix build with solr binari
...
+    step "Then the investigator gets a notification about the change"
+    step "and the notification is removed"
+    nl <- runReaderT
+        (do
+            -- TODO: repeats implementation of viewNotifications
+            nl <- lookupUserNotifications     
+            _ <- changeNotificationsDelivery (filter ((DeliverySite >) . notificationDelivered) nl) DeliverySite
+            -- nlAfter <- lookupUserNotifications -- TODO: ensure only delivered once
+            pure nl)
+        aiCtxt2
```
Comment (chreekat) [inline test/ModelTest.hs]: Perhaps these runReaderTs would be nicer if we defined `runFooContext = flip runReaderT fooCtxt` whenever we defined the ctxt. Or skip defining the ctxt and jump straight to defining the runner.

Comment (chreekat) [reply test/ModelTest.hs]: Or `runAsFoo`

File: src/Controller/Upload.hs [hunk 11]
```diff
@@ -91,69 +101,51 @@ chunkForm = do
...
   rb <- peeks Wai.requestBody
-  -- liftIO $ putStrLn "request body length"
-  -- liftIO $ print . BS.length =<< rb
-  n <- liftIO $ bracket
+  n <- liftIO (writeChunk off len file rb)
+  checkLength n -- TODO: clear block (maybe wait for calloc)
+  return $ emptyResponse noContent204 []
+
+writeChunk :: Int64 -> Word64 -> RawFilePath -> IO BS.ByteString -> IO Word64
+writeChunk off len file rb = bracket
```
[the remark this answers is not in the document]
Comment (kanishka-azimi) [reply src/Controller/Upload.hs]: I want to break off two pieces from this:
1. general file utility functions, as you note
2. the parts implementing the flow-js protocol

File: src/Service/Mail.hs [hunk 12]
```diff
@@ -24,7 +26,36 @@ import Has
...
+newtype MailPort = MailPort Int
+
+-- | Acount username for smtp mail delivery
+newtype MailUser = MailUser String
+
+-- | Account password for smtp mail delivery
+newtype MailPass = MailPass String
+
+-- | Parts of mailer that can be mocked
+data Mailer = Mailer
```
[the remark this answers is not in the document]
Comment (kanishka-azimi) [reply src/Service/Mail.hs]: I want it to have more fields later

File: test/Model/Factories.hs [hunk 13]
```diff
@@ -443,6 +446,20 @@ genVolumeLink =
...
+mkIngestInput volName =
+    object
+        [ ("name", String volName)
+        , ("containers"
+          , Array
+              (V.fromList
+                 [object
+                    [ ("name", "cont1")
+                    , ("key", "key1")
+                    , ("records", Array (V.fromList []))
```
[the remark this answers is not in the document]
Comment (kanishka-azimi) [reply test/Model/Factories.hs]: This module is too huge. we can enable when the Factories and ModelTest break up into smaller modules.

File: test/ModelTest.hs [hunk 14]
```diff
@@ -614,35 +626,149 @@ test_16 = ignoreTest $ -- TODO: enable this inside of nix build with solr binari
...
-test_17 = localOption (mkTimeout (10 * 10^(6 :: Int))) $ Test.stepsWithResourceAndTransaction "test_17" $ \step ist cn2 -> do
+test_register_volume_with_ezid :: TestTree
+test_register_volume_with_ezid = localOption (mkTimeout (15 * 10^(6 :: Int))) $
+    Test.stepsWithResourceAndTransaction "" $ \step ist cn2 -> do
+        step "Given an authorized investigator"
+        (aiAcct, aiCtxt) <- addAuthorizedInvestigatorWithInstitution' cn2
+        step "When the AI creates a partially shared volume"
+        step "and add data that will be indexed with ezid"
+        step "and the ezid generation runs"
+        -- TODO: should be lookup auth on rootParty
```
[the remark this answers is not in the document]
Comment (kanishka-azimi) [reply test/ModelTest.hs]: we need your createVolume helper sooner (:

File: test/ModelTest.hs (hunk 10 above)
Comment (kanishka-azimi) [reply test/ModelTest.hs]: > Or skip defining the ctxt and jump straight to defining the runner.
Explain

Comment (kanishka-azimi) [reply test/ModelTest.hs]: I get it actually, disregard last note.

Comment (chreekat) [reply test/ModelTest.hs]:
```haskell
-- current style
do
    noIdentNotifyCtxt <-
        (\n m l ml -> (mkDbContext cn2) {
                ctxNotifications = Just n
                , ctxMessages = Just m
                , ctxLogs = Just l
                , ctxMailer = Just ml
                })
            <$> mkNotificationsStub
            <*> loadMessages
            <*> mkLogsStub
            <*> pure mailer
    ...
    runReaderT
        (do
            stuff
            other things)
        noIdentNotifyCtxt

-- versus

do
    runNotifyWithNoIdent <-
        (\n m l ml -> flip runReaderT ((mkDbContext cn2) {
                ctxNotifications = Just n
                , ctxMessages = Just m
                , ctxLogs = Just l
                , ctxMailer = Just ml
                }))
            <$> mkNotificationsStub
            <*> loadMessages
            <*> mkLogsStub
            <*> pure mailer
    ...
    runNotifyWithNoIdent $ do
        stuff
        other things
```

**Comment by kanishka-azimi** [comment]
```
    simple notification:                         Retrieving mail config...
FAIL
      Exception: config/email: openBinaryFile: does not exist (No such file or directory)
```

**Pull request merged**